src/test/app/TheoreticalQuality_test.cpp

```diff
@@ -0,0 +1,568 @@
+        Env env(*this, supported_amendments());
+        for (int i = 0; i < numTestIterations; ++i)
+        {
+            auto const iterAsStr = to_string(i);
+            // New set of accounts on every iteration so the environment doesn't
+            // need to be recreated (2x speedup)
+            auto const alice = Account("alice" + iterAsStr);
+            auto const bob = Account("bob" + iterAsStr);
+            auto const carol = Account("carol" + iterAsStr);
+            auto const dan = Account("dan" + iterAsStr);
+            std::array<Account, numAccounts> accounts{alice, bob, carol, dan};
+            static_assert(
+                numAccounts == 4, "Path is only correct for four accounts");
+            path const accountsPath(accounts[1], accounts[2]);
+            env.fund(XRP(10000), alice, bob, carol, dan);
+            env.close();
+
+            // iterate through all pairs of accounts, randomly set the transfer
+            // rate, qIn, qOut, and if the account issues or redeems
+            for (std::size_t i = 0; i < numAccounts; ++i)
+            {
+                rndAccParams.maybeSetTransferRate(env, accounts[i]);
+                // The payment is from:
+                // account[0] -> account[1] -> account[2] -> account[3]
+                // set the trust lines and initial balances for each pair of
+                // neighboring accounts
+                std::size_t const j = i + 1;
+                if (j == numAccounts)
+                    continue;
+
+                rndAccParams.setupTrustLines(
+                    env, accounts[i], accounts[j], currency);
+                rndAccParams.maybeSetInitialBalance(
+                    env, accounts[i], accounts[j], currency);
+            }
+
+            // Accounts are set up, make the payment
+            IOU const iou{accounts.back(), currency};
+            RippleCalcTestParams rcp{env.json(
+                pay(accounts.front(), accounts.back(), iou(paymentAmount)),
+                accountsPath,
+                txflags(tfNoRippleDirect))};
+
+            testCase(rcp, env.closed());
+        }
+    }
+
+    void
+    testBookStep(boost::optional<int> const& reqNumIterations)
+    {
+        testcase("Book Step");
+        using namespace jtx;
+
+        // clang-format off
+
+        // Setup a payment through an offer: alice (USD/bob) -> bob -> (USD/bob)|(EUR/carol) -> carol -> dan
+        // For each relavent trust line, vary input quality, output quality, debt direction.
+        // For each account, vary transfer rate.
+        // The USD/bob|EUR/carol offer owner is "Oscar"
+
+        // clang-format on
+
+        int const numTestIterations = reqNumIterations.value_or(200);
+
+        constexpr std::uint32_t paymentAmount = 1;
+
+        Currency const eurCurrency = to_currency("EUR");
+        Currency const usdCurrency = to_currency("USD");
+
+        // Class to randommly set account transfer rates, qualities, and other
```

Comment:
randommly -> randomly?

Reply:
Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`